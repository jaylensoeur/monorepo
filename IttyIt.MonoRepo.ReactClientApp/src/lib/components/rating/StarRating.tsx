import * as React from "react";
import { StarTypeProps, StarRatingProps }  from './StarTypeProps';

const StarRating = ({rating}: StarRatingProps): JSX.Element => {
    const starRating = {
        half: "◐",
        full: "●",
        empty: "○",
    };
    const star: JSX.Element[] = [];

    for (let i = 0; i < 5; i++) {
        let starType: StarTypeProps = {};

        if (i < rating) {
            if (Math.floor(rating) === i && rating % 1) {
                starType.shape = starRating.half;
            } else {
                starType.shape = starRating.full;
            }
            starType.color = '#f1e50e';
            starType.className = 'star-rating star-rating--star';
        } else {
            starType = {
                className: 'star-rating star-rating--star',
                shape: starRating.empty,
                color: '#ccc'
            };
        }
        star.push(<span key={i} style={{color: starType.color}} className={starType.className}>{starType.shape}</span>);
    }

    return (
        <div data-testid="rating" className="rating-block">
            {star}
        </div>
    );
};

export default StarRating;

