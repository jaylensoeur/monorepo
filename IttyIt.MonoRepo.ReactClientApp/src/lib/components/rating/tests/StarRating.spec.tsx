import * as React from "react";
import { create } from "react-test-renderer";
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'

import StarRating from "../StarRating";

describe("<StarRating />", () => {
    it("should render", async () => {
        const component = create(<StarRating rating={1} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("should see 2 1/2 stars", async () => {
        render(<StarRating rating={2.5}/>);
        const ratingSection = await screen.getByTestId('rating');
        expect(ratingSection).toHaveTextContent('●●◐○○');
    });

    [
        {rating: 1, expected: '●○○○○'},
        {rating: 2, expected: '●●○○○'},
        {rating: 3, expected: '●●●○○'},
        {rating: 4, expected: '●●●●○'},
        {rating: 5, expected: '●●●●●'},
        {rating: 6, expected: '●●●●●'},
    ].map((dataProvider) => {
        it(`should see ${dataProvider.expected} when rating is ${dataProvider.rating} stars`, async () => {
            render(<StarRating rating={dataProvider.rating}/>);
            const ratingSection = await screen.getByTestId('rating');
            expect(ratingSection).toHaveTextContent(dataProvider.expected);
        });
    });

    [
        {rating: -1, expected: '○○○○○'},
        {rating: 2.5, expected: '●●◐○○'},
        {rating: 3.5, expected: '●●●◐○'},
        {rating: 4.5, expected: '●●●●◐'},
        {rating: 5.5, expected: '●●●●●'},
    ].map((dataProvider) => {
        it(`should see ${dataProvider.expected} when rating is ${dataProvider.rating} stars`, async () => {
            render(<StarRating rating={dataProvider.rating}/>);
            const ratingSection = await screen.getByTestId('rating');
            expect(ratingSection).toHaveTextContent(dataProvider.expected);
        });
    });
});
