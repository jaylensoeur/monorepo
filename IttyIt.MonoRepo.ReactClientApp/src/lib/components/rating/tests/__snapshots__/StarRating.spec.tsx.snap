// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarRating /> should render 1`] = `
<div
  className="rating-block"
  data-testid="rating"
>
  <span
    className="star-rating star-rating--star"
    style={
      Object {
        "color": "#f1e50e",
      }
    }
  >
    ●
  </span>
  <span
    className="star-rating star-rating--star"
    style={
      Object {
        "color": "#ccc",
      }
    }
  >
    ○
  </span>
  <span
    className="star-rating star-rating--star"
    style={
      Object {
        "color": "#ccc",
      }
    }
  >
    ○
  </span>
  <span
    className="star-rating star-rating--star"
    style={
      Object {
        "color": "#ccc",
      }
    }
  >
    ○
  </span>
  <span
    className="star-rating star-rating--star"
    style={
      Object {
        "color": "#ccc",
      }
    }
  >
    ○
  </span>
</div>
`;
